java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java

```diff
@@ -0,0 +1,876 @@
+    UNDEFINED(0),
+    /** COO sparse tensor. */
+    COO(1),
+    /** CSR or CSC sparse tensor. */
+    CSRC(2),
+    /** Block sparse tensor. */
+    BLOCK_SPARSE(4);
+
+    /** The int value mirroring OrtSparseFormat. */
+    public final int value;
+
+    private static final SparseTensorType[] values = new SparseTensorType[5];
+
+    static {
+      values[0] = UNDEFINED;
+      values[1] = COO;
+      values[2] = CSRC;
+      values[3] = UNDEFINED;
+      values[4] = BLOCK_SPARSE;
+    }
+
+    SparseTensorType(int value) {
+      this.value = value;
+    }
+
+    /**
+     * Maps from an int in native land into a SparseTensorType instance.
+     *
+     * @param value The value to lookup.
+     * @return The enum instance.
+     */
+    public static SparseTensorType mapFromInt(int value) {
+      if ((value > 0) && (value < values.length)) {
+        return values[value];
+      } else {
+        return UNDEFINED;
+      }
+    }
+  }
+
+  /**
+   * Top level interfaces for the Java side representation of a sparse tensor.
+   *
+   * <p>Will be sealed to {@link COOTensor}, {@link CSRCTensor} and {@link BlockSparseTensor} when
+   * possible.
+   *
+   * <p>Does not support Strings as a data type.
+   *
+   * @param <T> The indices buffer type.
+   */
+  public static interface SparseTensor<T extends Buffer> {
+    /**
+     * Gets the dense shape of the sparse tensor.
+     *
+     * @return The sparse tensor shape.
+     */
+    public long[] getShape();
+
+    /**
+     * Gets the shape of the values of the sparse tensor.
+     *
+     * @return The sparse tensor value shape.
+     */
+    public long[] getValuesShape();
+
+    /**
+     * Gets the shape of the indices of the sparse tensor.
+     *
+     * @return The sparse tensor indices shape.
+     */
+    public long[] getIndicesShape();
+
+    /**
+     * The data type of the sparse tensor.
+     *
+     * @return The sparse tensor data type.
+     */
+    public OnnxJavaType getType();
+
+    /**
+     * The sparsity type of the sparse tensor.
+     *
+     * @return The sparse tensor sparsity type.
+     */
+    public SparseTensorType getSparsityType();
+
+    /**
+     * The number of non-zero elements.
+     *
+     * @return The number of non-zero elements.
+     */
+    public long getNumNonZeroElements();
+
+    /**
+     * The indices type of the sparse tensor.
+     *
+     * <p>Only {@link OnnxJavaType#INT32} and {@link OnnxJavaType#INT64} are supported.
+     *
+     * @return The sparse tensor indices type.
+     */
+    public OnnxJavaType getIndicesType();
+
+    /**
+     * Get the indices buffer.
+     *
+     * @return The indices buffer.
+     */
+    public T getIndices();
+
+    /**
+     * Get the data buffer.
+     *
+     * @return The data buffer.
+     */
+    public Buffer getData();
+  }
+
+  /** Abstract base class for Java sparse tensors */
+  private abstract static class BaseSparseTensor<T extends Buffer> implements SparseTensor<T> {
```

Comment:
>BaseSparseTensor [](http://example.com/codeflow?start=32&length=16)

Just curious, why do we need both base class (which is not used fully IMHO) and the interface? Can base class serve as an interface?

Reply:
The interface + abstract base class is a fairly standard idiom in Java, though I guess it doesn't offer much flexibility here given I intend to seal the interface so it only has the provided implementations. I can refactor it into a public abstract class.

Reply:
`SparseTensor` and `OnnxSparseTensor` represent different objects with different uses. The `SparseTensor` is purely Java side, it holds no references to ORT objects or data and is used as the way to prepare data in Java for supplying into ORT, or for working on the output if a sparse tensor is produced from a run call. `SparseTensor` and its implementations could easily live outside the `OnnxSparseTensor` class, and if Java had a similar package to scipy which provided sparse tensors it wouldn't exist at all we'd just depend on the external definitions. `OnnxSparseTensor` holds the references to the ORT objects and memory, but the data only be accessed from Java by making a `SparseTensor` and copying the data out because there aren't any direct accessors into the ORT sparse tensor object. The `OnnxSparseTensor` has a reference to the buffers that came from the `SparseTensor` to make sure they live longer than the `OnnxSparseTensor`, though I'm not actually sure now if it's copying the memory into the `OrtValue` or referencing it.

`OnnxTensorLike` is a made up class which exists because both `OnnxTensor` and `OnnxSparseTensor` can be inputs to a run call, but `OnnxSequence` and `OnnxMap` cannot. So I need something in the type hierarchy that lets me distinguish them statically.

Reply:
Ooops, no wait, you're right, I'm missing the copy in `OnnxSparseTensor.getDataBuffer()`, `OnnxSparseTensor.getIndicesBuffer()` and `OnnxSparseTensor.getInnerIndicesBuffer()`. I'll fix that.

Reply:
I've fixed the get methods so they copy the buffer. Once this has merged in I plan to add methods to both `OnnxSparseTensor` and `OnnxTensor` which let users copy the buffer into one they supply, and hopefully that will let me refactor and reduce the get logic a little (by pushing it into `OnnxTensorLike`).